### PR TITLE
add support to batch metric and send on a strict interval

### PR DESCRIPTION
**Pull request by maciuszek**
Add more definite send interval control than what flush interval provided, restricting when any metrics is sent to this interval. This will implicitly introduce batching for timers.

### CODE REVIEW

File: net_sink.go [hunk 1]
```diff
@@ -312,27 +317,41 @@ func (s *netSink) run() {
...
+		// if the channel is at capacity, it's blocking, let's signal for it to be flushed and unblock
+		if len(s.outc) == cap(s.outc) {
+			s.Flush()
+		}
+
 		select {
-		case <-t.C:
-			s.flush()
 		case done := <-s.doFlush:
-			// Only flush pending buffers, this prevents an issue where
```
Comment (maciuszek) [inline net_sink.go:319]: this case was effectively redundant, competing with `case buf := <-s.outc` to process and clear outc . repurposing it to handle batching

**Review by sokada1221 — COMMENTED** [review]
A few non-binding comments - as I may not have full context/understanding:
* I think it's generally a good idea to implement batching with 2 parameters: batch size and interval. Since interval is already configurable, it'd be nice to have batch size configurable - with default being the magic `approxMaxMemBytes/bufSize` lol
* How do you plan to test the change? Should we add some metrics/logs to verify the batching behavior?

File: stats_test.go [hunk 2]
```diff
@@ -89,10 +89,11 @@ func TestValidateTags(t *testing.T) {
 	store.NewCounterWithTags("test", map[string]string{"host": "i"}).Inc()
 	store.Flush()
 
-	expected = "reserved_tag:1|c\ntest.__host=i:1|c"
-	counter = sink.record
-	if !strings.Contains(counter, expected) {
-		t.Error("wanted counter value of test.___f=i:1|c, got", counter)
+	expected = "test.__host=i:1|c"
```
Comment (maciuszek) [inline stats_test.go:92]: Note: this test is out of scope of this work but it was previously volatile with the order of reserved_tag vs test.__host not being deterministic

File: net_sink.go [hunk 3]
```diff
@@ -118,8 +118,8 @@ func NewNetSink(opts ...SinkOption) FlushableSink {
 		bufSize = defaultBufferSizeTCP
 	}
 
-	s.outc = make(chan *bytes.Buffer, approxMaxMemBytes/bufSize)
-	s.retryc = make(chan *bytes.Buffer, 1) // It should be okay to limit this given we preferentially process from this over outc.
+	s.outc = make(chan *bytes.Buffer, approxMaxMemBytes/bufSize) // todo: need to understand why/how this number was chosen and probably elevate it
```
Comment (maciuszek) [inline net_sink.go]: ~To make threading possible forking https://github.com/lyft/gostats/pull/169#pullrequestreview-2521121502 to here . @sokada1221.~

~So this doesn't restrict memory allocated, but the amount of slots per metric/string. 
If it exceeds it'll block more stats for being written until read, it wouldn't act as a batching mechanism 🤔. buffered channeled are a bit strange, i think in actuality this buffer will always be full, if we would change it to normal channel with no buffer (always block), i don't think we would see an impact.~

File: net_sink.go [hunk 4]
```diff
@@ -312,34 +319,51 @@ func (s *netSink) run() {
...
+func (s *netSink) sendBatch(batch []bytes.Buffer) []bytes.Buffer {
+	n := len(batch)
+	for i := 0; i < n && s.conn != nil; i++ {
+		buf := batch[i]
+		// todo can we just send all the batches in one call?
+		if err := s.send(&buf); err == nil {
+			putBuffer(&buf)
 		}
 	}
+	return batch[:0]
```
Comment (charlievieth) [inline net_sink.go]: If there is an error in `send` this will cause any following batched metrics to be dropped. Might be better to keep them queued. Additionally, this prevents us from putting any buffers received after the error back into the buffer pool (which is only a problem if we decide to keep the existing logic).

Comment (maciuszek) [reply net_sink.go]: ~If there's an error in send, the metrics will be written to the retryc channel so nothing should be dropped i think, despite clearing it from the batch. In the current state, that retry handling will escape batching altogether, will think on this.~

Nevermind, I see any failure prevents any/all subsequent sends in the current iteration, good catch, thanks

File: settings.go [hunk 5]
```diff
@@ -20,6 +20,8 @@ const (
 	DefaultFlushIntervalS = 5
 	// DefaultLoggingSinkDisabled is the default behavior of logging sink suppression, default is false.
 	DefaultLoggingSinkDisabled = false
+	// DefaultBatchSize is the default maximum amount of stats we batch before sending, default is 0 which disables batching.
```
Comment (paulinjo) [inline settings.go]: So we're batching based on a number of stats here instead of the size in MB? In that case I would think a batch of 1 would be equivalent to no batching, not 0; a 0 batch size doesn't really make sense.

But even better if we could explicitly enable/disable batching via config, I think that would make things more clear.

Comment (maciuszek) [reply settings.go]: Exactly yeah, the batch size specifies the number of stats we batch.

batch size 1 will effectively be similar as no batching, but 0 specifically reverts back to the old behaviour hence my use of it, as in don't go through the batch code at all.  

We could have 2 configurations for this, since i'm effectively using 0 like a feature flag anyway. But will we really ever clean it up, and having 1 configuration hard dependant on the other isn't too great.

Comment (maciuszek) [reply settings.go]: Done. Made the configuration more granular

File: net_sink.go [hunk 6]
```diff
@@ -277,10 +278,26 @@ func (s *netSink) run() {
...
 
+	isBatchEnabled := s.conf.BatchEnabled
+
+	batchSize := s.conf.BatchSize
+	batch := make([]bytes.Buffer, 0, batchSize+cap(s.outc)) // overallocate to consider draining outstanding outc data. despite the expanded allocation, batchSize is still used to determine if we send the batched stats
+
+	batchTimeout := time.Duration(s.conf.BatchSendIntervalS) * time.Second
+	batchInterval := time.After(batchTimeout)
+
+	sendBatch := false
```
Comment (paulinjo) [inline net_sink.go]: nit: rename to `doSendBatch` or `shouldSendBatch`

File: net_sink.go [hunk 7]
```diff
@@ -307,39 +324,100 @@ func (s *netSink) run() {
...
 			}
 			putBuffer(buf)
-			continue
+			continue // we will need to reconnect if an error occurs, so go back to the top of the next iteration to see
 		default:
 			// Drop through in case retryc has nothing.
 		}
 
+		// send batched outc data anytime indicated or the batch is full
+		if sendBatch || len(batch) >= batchSize {
```
Comment (paulinjo) [inline net_sink.go]: The amount of nesting and conditions here is a bit hard to follow. Hopefully there's some way to refactor that would improve readability.

Comment (maciuszek) [reply net_sink.go]: Refactored the code a bit and I think it's a bit more readable.

The only alternative I can think of is forking another run function for batching and determine which to use when launching the go routine.

Doing that may indeed be more efficient since there will be less conditions per loop, but the drawback is we'll have duplicate code which will be hard to maintain in 2 places 🤔 

What are your thoughts, I could modularize some of the internals but the structure will effectively be duplicated.

**Review by paulinjo — DISMISSED** [review]
